src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,535 @@
+        return torch.tensor(ids), torch.tensor(vectors)
+
+
+class HFIndex(Index):
+    """
+    A wrapper around an instance of :class:`~datasets.Datasets`. If ``index_path`` is set to ``None``,
+    we load the pre-computed index available with the :class:`~datasets.arrow_dataset.Dataset`, otherwise, we load the index from the indicated path on disk.
+
+    Args:
+        dataset (:obj:`str`, optional, defaults to ``wiki_dpr``):
+            A datatset identifier of the indexed dataset on HuggingFace AWS bucket (list all available datasets and ids with ``datasets.list_datasets()``).
+        dataset_split (:obj:`str`, optional, defaults to ``train``)
+            Which split of the ``dataset`` to load.
+        index_name (:obj:`str`, optional, defaults to ``train``)
+            The index_name of the index associated with the ``dataset``. The index loaded from ``index_path`` will be saved under this name.
+        index_path (:obj:`str`, optional, defaults to ``None``)
+            The path to the serialized faiss index on disk.
+    """
+
+    def __init__(
+        self,
+        dataset,
+        dataset_split,
+        index_name,
+        index_path,
+        use_dummy_dataset,
+    ):
+        super().__init__()
+        self.dataset = dataset
+        self.dataset_split = dataset_split
+        self.index_name = index_name
+        self.index_path = index_path
+        self._index_initialize = False
+        self.use_dummy_dataset = use_dummy_dataset
+
+        logger.info("Loading passages from {}".format(self.dataset))
+        self.index = load_dataset(
+            self.dataset, with_index=False, split=self.dataset_split, dummy=self.use_dummy_dataset
+        )
+
+    def is_initialized(self):
+        return self._index_initialize
+
+    def init_index(self):
+        if self.index_path is not None:
+            logger.info("Loading index from {}".format(self.index_path))
+            self.index.load_faiss_index(index_name=self.index_name, file=self.index_path)
+        else:
+            logger.info("Loading index from {}".format(self.dataset + " with index name " + self.index_name))
+            self.index = load_dataset(
+                self.dataset,
+                with_embeddings=True,
+                with_index=True,
+                split=self.dataset_split,
+                index_name=self.index_name,
+                dummy=self.use_dummy_dataset,
+            )
+        self._index_initialize = True
+
+    def get_doc_dicts(self, doc_ids):
+        return [self.index[doc_ids[i].tolist()] for i in range(doc_ids.shape[0])]
+
+    def get_top_docs(self, question_hidden_states, n_docs=5):
+        _, docs = self.index.get_nearest_examples_batch("embeddings", question_hidden_states, n_docs)
+        ids = [[int(i) for i in doc["id"]] for doc in docs]
+        vectors = [doc["embeddings"] for doc in docs]
+        return torch.tensor(ids), torch.tensor(vectors)
```

Comment:
In the future we might want the retriever to output pt/tf/np tensors like the tokenizers do (can be kept in a follow-up PR).